src/Workspaces/CoreTest/SolutionTests/TryApplyChangesTests.cs

```diff
@@ -0,0 +1,171 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.UnitTests
+{
+    [UseExportProvider]
+    public class TryApplyChangesTests
+    {
+        private class CustomizedCanApplyWorkspace : Workspace
+        {
+            private readonly ImmutableArray<ApplyChangesKind> _allowedKinds;
+            private readonly Func<ParseOptions, ParseOptions, bool>? _canApplyParseOptions;
+            private readonly Func<CompilationOptions, CompilationOptions, bool>? _canApplyCompilationOptions;
+
+            public CustomizedCanApplyWorkspace(params ApplyChangesKind[] allowedKinds)
+                : this(allowedKinds, canApplyParseOptions: null)
+            {
+            }
+
+            public CustomizedCanApplyWorkspace(ApplyChangesKind[] allowedKinds,
+                Func<ParseOptions, ParseOptions, bool>? canApplyParseOptions = null,
+                Func<CompilationOptions, CompilationOptions, bool>? canApplyCompilationOptions = null)
+                : base(Host.Mef.MefHostServices.DefaultHost, workspaceKind: nameof(CustomizedCanApplyWorkspace))
+            {
+                _allowedKinds = allowedKinds.ToImmutableArray();
+                _canApplyParseOptions = canApplyParseOptions;
+                _canApplyCompilationOptions = canApplyCompilationOptions;
+
+                // Add a C# project automatically so each test has something to try mutating
+                OnProjectAdded(ProjectInfo.Create(ProjectId.CreateNewId(), VersionStamp.Default, "TestProject", "TestProject", LanguageNames.CSharp));
+            }
+
+            public override bool CanApplyChange(ApplyChangesKind feature)
+            {
+                return _allowedKinds.Contains(feature);
+            }
+
+            public override bool CanApplyParseOptionChange(ParseOptions oldOptions, ParseOptions newOptions, Project project)
+            {
+                if (_canApplyParseOptions != null)
+                {
+                    return _canApplyParseOptions(oldOptions, newOptions);
+                }
+
+                return base.CanApplyParseOptionChange(oldOptions, newOptions, project);
```

Comment:
```suggestion
                    return _canApplyParseOptions?.Invoke(oldOptions, newOptions) ?? base.CanApplyParseOptionChange(oldOptions, newOptions, project);
```

(up to you though).

Reply:
Not really sure that's more readable, although it is cute...

Reply:
(going to leave it as is)